src/pgwire/src/protocol.rs

```diff
@@ -407,20 +425,33 @@ where
     }
 
     async fn advance_drain(&mut self) -> Result<State, io::Error> {
-        match self.conn.recv().await? {
+        let message = self.conn.recv().await?;
+        if message.is_some() {
+            self.adapter_client()
+                .remove_idle_in_transaction_session_timeout();
+        }
+        match message {
             Some(FrontendMessage::Sync) => self.sync().await,
             None => Ok(State::Done),
             _ => Ok(State::Drain),
         }
     }
 
+    async fn advance_timed_out(&mut self, err: ErrorResponse) -> Result<State, io::Error> {
+        // dropping the adapter client will terminate the session.
+        self.adapter_client = None;
+        let _ = self.conn.recv().await?;
```

Comment:
why does the recv need to exist?

Reply:
This is to mimic the PostgreSQL behavior. PostgreSQL will not send an error until the client tries to send another query. I believe it's because the wire protocol doesn't allow PG to send error responses without first receiving send type of request message.

Reply:
Good point. Can you add a comment about that?

Reply:
Just added a comment.

Reply:
Just added a comment.